From 7c16fb3e412b5d4245bf0134f482f8441c9b9731 Mon Sep 17 00:00:00 2001
From: "qmai@libertyglobal.com" <qmai@libertyglobal.com>
Date: Thu, 22 Apr 2021 20:37:44 +0000
Subject: [PATCH] send Renew after T1 timer expires ( OFW 1143 )

Previously the DHCPv6 client would send a Solicit message after the
T1 timer expires. Send a Renew instead.

v2
---
 ClntMessages/ClntMsgRequest.cpp | 11 +++++++++--
 ClntMessages/ClntMsgSolicit.cpp |  7 +++++++
 2 files changed, 16 insertions(+), 2 deletions(-)

diff --git a/ClntMessages/ClntMsgRequest.cpp b/ClntMessages/ClntMsgRequest.cpp
index 9ae14132..18f4acb2 100644
--- a/ClntMessages/ClntMsgRequest.cpp
+++ b/ClntMessages/ClntMsgRequest.cpp
@@ -62,8 +62,9 @@ TClntMsgRequest::TClntMsgRequest(TOptList opts, int iface)
     SPtr<TClntMsg> advertise = SPtr_cast<TClntMsg>(ClntTransMgr().getAdvertise());
     copyAAASPI(SPtr_cast<TClntMsg>(advertise));
 
-    // remove just used server
-    ClntTransMgr().delFirstAdvertise();
+    // There will be no server available for REQUEST.
+    // remove just used server later.
+    // ClntTransMgr().delFirstAdvertise();
 
     // copy whole list from SOLICIT ...
     Options = opts;
@@ -205,6 +206,12 @@ void TClntMsgRequest::doDuties()
     // timeout is reached and we still don't have answer, retransmit
     if (RC>MRC) 
     {
+        // remove all used servers from the list
+        while (ClntTransMgr().getAdvertiseLstCount())
+        {
+            ClntTransMgr().delFirstAdvertise();
+        }
+
         ClntTransMgr().sendRequest(Options, Iface);
 
         IsDone = true;
diff --git a/ClntMessages/ClntMsgSolicit.cpp b/ClntMessages/ClntMsgSolicit.cpp
index 06be83f0..67af02ff 100644
--- a/ClntMessages/ClntMsgSolicit.cpp
+++ b/ClntMessages/ClntMsgSolicit.cpp
@@ -109,6 +109,13 @@ TClntMsgSolicit::TClntMsgSolicit(int iface, SPtr<TIPv6Addr> addr,
     appendAuthenticationOption();
     
     IsDone = false;
+
+    // remove all used servers from the list
+    while (ClntTransMgr().getAdvertiseLstCount())
+    {
+        ClntTransMgr().delFirstAdvertise();
+    }
+
     send();
 }
 
-- 
2.24.0

